src/ada_data_url.cpp

```diff
@@ -0,0 +1,132 @@
+#include <string_view>
+#include <cctype>
+
+#include "ada.h"
+
+namespace ada::data_url {
+
+ada::data_url::data_url parse_data_url(std::string_view data_url) {
+  auto out = ada::data_url::data_url();
+
+  auto url = ada::parse<ada::url>(data_url, nullptr);
+
+  // 1. Assert: dataURL’s scheme is "data".
+  if (!url || url->get_protocol() != "data:") {
+      out.is_valid = false;
+      return out;
+  }
+
+  // 2. Let input be the result of running the URL serializer on dataURL with exclude
+  //    fragment set to true.
+  url->set_hash({});
+  auto input = url->get_href();
+
+  // 3. Remove the leading "data:" from input.
+  input.erase(0, 5);
+
+  // 4. Let position point at the start of input.
+  size_t position = 0;
+
+  // 5. Let mimeType be the result of collecting a sequence of code points that are
+  //    not equal to U+002C (,), given position.
+  auto mimetype = collect_sequence_of_code_points(',', input, position);
+  auto mimetype_length = mimetype.length();
+
+  // 6. Strip leading and trailing ASCII whitespace from mimeType.
+  mimetype = removeASCIIWhiteSpace(mimetype, true, true);

```

Comment:
Is the removal of white space needed, do we have an example where it is needed?

Reply:
It is needed, it will be tested once I port over the WPTs.